### PR TITLE
GODRIVER-1494 add option for encoding/json style map encoding

### PR DESCRIPTION
**Pull request by iwysiu**

### CODE REVIEW

File: bson/bsoncodec/map_codec.go [hunk 1]
```diff
@@ -20,8 +22,9 @@ var defaultMapCodec = NewMapCodec()
 
 // MapCodec is the Codec used for map values.
 type MapCodec struct {
-	DecodeZerosMap   bool
-	EncodeNilAsEmpty bool
+	DecodeZerosMap         bool
+	EncodeNilAsEmpty       bool
+	EncodeKeysWithStringer bool
```
Comment (craiggwilson) [inline bson/bsoncodec/map_codec.go]: As a thought, what if you made this a function instead and let the user provide you with something to convert when they want it different than the standard?

```
KeyEncoder func(val reflect.Value) (string, error)
```

You can provide a function that does this already, so it's just as easy as setting true.

File: bson/bson_test.go [hunk 2]
```diff
@@ -112,6 +117,37 @@ func TestD(t *testing.T) {
...
 }
 
+type stringerString string
+
+func (ss stringerString) String() string {
+	return "bar"
+}
+
+func TestMapCodec(t *testing.T) {
+	t.Run("EncodeKeysWithStringer", func(t *testing.T) {
```
Comment (craiggwilson) [inline bson/bson_test.go]: So, what about decoding... My assertion here is that this doesn't work because when we read it back we get "bar", which is decidedly _not_ "foo".  

In the ticket, this isn't a problem because they are encoding them as the underlying string value, not the result of the String() function. But if the type is an integer, we'll be writing it as a string and then decoding it as a string to assign it to an int? Do we handle that already?

Anyways, perhaps you need a 
```
type KeyCodec interface {
  Encode(val reflect.Value) (string, error)
  Decode(val string) (reflect.Value, error)
}
```

File: bson/bsoncodec/map_codec.go (hunk 1 above)
Comment (craiggwilson) [inline bson/bsoncodec/map_codec.go]: Also, doesn't this defaulting to false change the default behavior?

File: bson/bson_test.go (hunk 2 above)
Comment (iwysiu) [reply bson/bson_test.go]: As discussed, the default behavior was switched to the encoding/json behavior, which can be roundtripped

File: bson/bsoncodec/map_codec.go (hunk 1 above)
Comment (iwysiu) [reply bson/bsoncodec/map_codec.go]: Due to how the packages are set up, there isn't a good way to add an interface option, so it stayed as a bool. I am open to other option names

Comment (iwysiu) [reply bson/bsoncodec/map_codec.go]: went away

File: bson/bsoncodec/map_codec.go [hunk 3]
```diff
@@ -22,6 +24,7 @@ var defaultMapCodec = NewMapCodec()
 type MapCodec struct {
 	DecodeZerosMap   bool
 	EncodeNilAsEmpty bool
+	MgoKeyHandling   bool
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go:27]: I'd prefer to contain any named references to `mgo` to the `mgocompat` package and have these options be generically named, but set appropriately to create the `mgocompat.Registry` instead. Can we rename this to something like `UseStringerForKeys`?

File: bson/bsoncodec/map_codec.go [hunk 4]
```diff
@@ -204,3 +209,95 @@ func clearMap(m reflect.Value) {
...
+			}
+			keyVal = reflect.ValueOf(n).Convert(keyType)
+		default:
+			return keyVal, errors.New("Unexpected key type")
+		}
+	}
+	return keyVal, err
+}
+
+type mgoKeyCodec struct{}
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go:276]: Where is this used?

File: bson/bsoncodec/map_codec.go [hunk 5]
```diff
@@ -79,7 +85,16 @@ func (mc *MapCodec) mapEncodeValue(ec EncodeContext, dw bsonrw.DocumentWriter, v
 
 	keys := val.MapKeys()
 	for _, key := range keys {
-		keyStr := fmt.Sprint(key)
+		var keyStr string
+		if mc.MgoKeyHandling {
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go:89]: I think we can consolidate the encode functions:

```
if mc.MgoKeyHandling {
    return fmt.Sprint(key)
}

if key is string {
    return key
}

// other logic for TextMarshaler and native types that we can convert

File: bson/bsonoptions/map_codec_options.go [hunk 6]
```diff
@@ -29,9 +30,15 @@ func (t *MapCodecOptions) SetEncodeNilAsEmpty(b bool) *MapCodecOptions {
...
 
+// SetMgoKeyHandling specifies if keys should be encoded and decoded with mgo's handling. Defaults to false.
+func (t *MapCodecOptions) SetMgoKeyHandling(b bool) *MapCodecOptions {
+	t.MgoKeyHandling = &b
+	return t
+}
+
 // MergeMapCodecOptions combines the given *MapCodecOptions into a single *MapCodecOptions in a last one wins fashion.
 func MergeMapCodecOptions(opts ...*MapCodecOptions) *MapCodecOptions {
-	s := MapCodec()
```
Comment (divjotarora) [inline bson/bsonoptions/map_codec_options.go:34]: Why this change?

File: bson/bsoncodec/map_codec.go (hunk 3 above)
Comment (craiggwilson) [reply bson/bsoncodec/map_codec.go:27]: Call it `EncodeKeysWithStringer` to maych the `EncodeNilAsEmpty`

I'm still wondering why we can't take a `func(reflect.Value) (string, error)` here...  `KeyEncoder` and then have a `KeyDecoder` as well.

File: bson/bsoncodec/map_codec.go [hunk 7]
```diff
@@ -204,3 +199,80 @@ func clearMap(m reflect.Value) {
...
 }
+
+func (mc *MapCodec) encodeKey(val reflect.Value) (string, error) {
+	if mc.MgoKeyHandling {
+		return fmt.Sprint(val), nil
+	}
+
+	// keys of any string type are used directly
+	if val.Kind() == reflect.String {
+		return val.String(), nil
```
Comment (liambrown89) [inline bson/bsoncodec/map_codec.go:210]: This has the same (or very similar) behaviour to `fmt.Sprint(val)` above - i.e. both use the stringer? Was this intended to be `string(val)` instead to use it directly rather than calling the `String`er?

File: bson/bsoncodec/map_codec.go (hunk 3 above)
Comment (divjotarora) [reply bson/bsoncodec/map_codec.go:27]: This keeps parity with our other interfaces (Marshaler/ValueMarshaler/Proxy/etc). I also prefer to avoid exposing the `reflect` package if we can. I know our overall codec system is built on it, but if we can get away with a simpler function, that's my preference. Also, if we changed this to explicitly take a function, this behavior would be opt-in rather than opt-out and using `Stringer` would be the default, which we've agreed is broken.

There is one remaining discussion: do we want to take `encoding.TextMarshaler` from the standard library's `encoding` package or make our own interface/function? Do we expect a user to want different behavior when using stdlib json vs our BSON library?

Comment (craiggwilson) [reply bson/bsoncodec/map_codec.go:27]: fair enough.

**Comment by iwysiu** [comment]
Divjot and I talked, and we decided that it was less likely to be backwards breaking for users if we use our own interfaces instead of encoding/json's

File: bson/bsoncodec/map_codec.go [hunk 8]
```diff
@@ -204,3 +213,80 @@ func clearMap(m reflect.Value) {
...
+	if mc.EncodeKeysWithStringer {
+		return fmt.Sprint(val), nil
+	}
+
+	// keys of any string type are used directly
+	if val.Kind() == reflect.String {
+		return val.String(), nil
+	}
+	// KeyMarshalers are marshaled
+	if tm, ok := val.Interface().(KeyMarshaler); ok {
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go:227]: nit: I assume `tm` was for `TextMarshaler`. Change to `km`.

Comment (divjotarora) [reply bson/bsoncodec/map_codec.go:227]: Should we be checking if a pointer to value implements `KeyMarshaler` as well? Our standard registry lookup checks that for other interfaces we support like `Marshaler`.

File: bson/bsoncodec/map_codec.go [hunk 9]
```diff
@@ -204,3 +213,80 @@ func clearMap(m reflect.Value) {
...
+}
+
+var keyUnmarshalerType = reflect.TypeOf((*KeyUnmarshaler)(nil)).Elem()
+
+func (mc *MapCodec) decodeKey(key string, keyType reflect.Type) (reflect.Value, error) {
+	keyVal := reflect.ValueOf(key)
+	var err error
+	switch {
+	case !mc.EncodeKeysWithStringer && reflect.PtrTo(keyType).Implements(keyUnmarshalerType):
+		keyVal = reflect.New(keyType)
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go:254]: nit: Add a comment explaining what's happening here. It took me a while to figure out how this works.

File: bson/bsoncodec/map_codec.go [hunk 10]
```diff
@@ -204,3 +213,80 @@ func clearMap(m reflect.Value) {
...
+		err = v.UnmarshalKey(key)
+		keyVal = keyVal.Elem()
+	default:
+		switch keyType.Kind() {
+		case reflect.String:
+			keyVal = reflect.ValueOf(key).Convert(keyType)
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			s := string(key)
+			n, parseErr := strconv.ParseInt(s, 10, 64)
+			if parseErr != nil || reflect.Zero(keyType).OverflowInt(n) {
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go:265]: I don't see these `Overflow*` checks in our existing code. Are they intentional? Also, the `ParseInt` function takes a `bitSize` which tells it the size the number should be (e.g. an `int8` should be 8 bits). It will return a custom error if the number can't fit in the specified bit size. Is there a reason we're always specifying `bitSize=64` rather than taking advantage of that functionality?

File: bson/bsoncodec/map_codec.go [hunk 11]
```diff
@@ -204,3 +213,80 @@ func clearMap(m reflect.Value) {
...
+				parsed, err := strconv.ParseFloat(key, 64)
+				if err != nil {
+					return keyVal, fmt.Errorf("Map key is defined to be a decimal type (%v) but got error %v", keyType.Kind(), err)
+				}
+				keyVal = reflect.ValueOf(parsed)
+				break
+			}
+			fallthrough
+		default:
+			return keyVal, errors.New("Unexpected key type")
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go:288]: Wrap the error instead of swallowing it: `fmt.Errorf("unsupported key type: %s")`

File: bson/bsonoptions/map_codec_options.go [hunk 12]
```diff
@@ -10,6 +10,12 @@ package bsonoptions
 type MapCodecOptions struct {
 	DecodeZerosMap   *bool // Specifies if the map should be zeroed before decoding into it. Defaults to false.
 	EncodeNilAsEmpty *bool // Specifies if a nil map should encode as an empty document instead of null. Defaults to false.
+	// Specifies how keys should be handled. If false, the behavior matches encoding/json, where the encoding key type must
+	// either be a string, an integer type, or implement encoding.TextMarshaler and the decoding key type must either be a
```
Comment (divjotarora) [inline bson/bsonoptions/map_codec_options.go:14]: Update the interface names to the `bsoncodec` interfaces.

File: bson/bsonoptions/map_codec_options.go [hunk 13]
```diff
@@ -23,12 +29,22 @@ func (t *MapCodecOptions) SetDecodeZerosMap(b bool) *MapCodecOptions {
...
 }
 
-// SetEncodeNilAsEmpty specifies  if a nil map should encode as an empty document instead of null. Defaults to false.
+// SetEncodeNilAsEmpty specifies if a nil map should encode as an empty document instead of null. Defaults to false.
 func (t *MapCodecOptions) SetEncodeNilAsEmpty(b bool) *MapCodecOptions {
 	t.EncodeNilAsEmpty = &b
 	return t
 }
 
+// SetEncodeKeysWithStringer specifies how keys should be handled. If false, the behavior matches encoding/json, where the
```
Comment (divjotarora) [inline bson/bsonoptions/map_codec_options.go:38]: Update the interface names here as well.

File: bson/bsoncodec/map_codec.go (hunk 8 above)
Comment (iwysiu) [reply bson/bsoncodec/map_codec.go]: from what i can tell, map keys aren't addressable, so I didn't any of the pointer to value things.

Comment (divjotarora) [reply bson/bsoncodec/map_codec.go]: Yep, seems like `encoding/json` doesn't respect &T implementing `TextMarshaler` either. As discussed offline, this is likely because map keys and values can move around if the map is re-sized so any addresses to them would be invalidated.
